core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -4223,6 +4243,114 @@ private void handleScalarSubQuery(
+    // It is because that in sql-select list, the SqlFunctionCategory is USER_DEFINED_FUNCTION
+    // for a SqlUnresolvedFunction,so we should do this change.
+    if (function instanceof SqlUnresolvedFunction) {
+      if (!function.getFunctionType().isTableFunction()) {
+        SqlFunction newFunction =
+            new SqlUnresolvedFunction(function.getNameAsId(),
+                function.getReturnTypeInference(),
+                function.getOperandTypeInference(),
+                function.getOperandTypeChecker(),
+                function.getParamTypes(),
+                SqlFunctionCategory.USER_DEFINED_TABLE_FUNCTION);
+        functionCall.setOperator(newFunction);
+        function = newFunction;
+      }
+    }
+    // Check functionCall whether is a table function
+    List<SqlOperator> overloads  = new ArrayList<>();
+    opTab.lookupOperatorOverloads(function.getNameAsId(),
+        function.getFunctionType(),
+        function.getSyntax(), overloads, catalogReader.nameMatcher());
+    if (overloads.size() == 0) {
+      throw newValidationError(functionCall,
+          RESOURCE.exceptTableFunction(function.getName()));
+    }
+    // Check the parent select whether is a aggregate statement
+    if (isAggregate(parentSelect)) {
+      throw newValidationError(functionCall,
+          RESOURCE.notAllowTableFunctionInAggregate());
+    }
+    SqlNodeList aliasNodes
+        = (SqlNodeList) selectItem.getOperands()[1];
+    List<String> aliasList = new ArrayList<>();
+    for (SqlNode aliasNode : aliasNodes) {
+      aliasList.add(deriveAlias(aliasNode, aliasList.size()));
+    }
+    aliases.addAll(aliasList);
+
+    String tableAlias = TABLE_FUNCTION_PREFIX + nextGeneratedId++;
+    // Expand the table function alias
+    for (String alias : aliasList) {
+      SqlIdentifier id = new SqlIdentifier(Lists.newArrayList
+          (tableAlias, alias), SqlParserPos.ZERO);
+      expandedSelectItems.add(id);
+    }
+
+    // Register namespace for table function
+    SqlValidatorScope fromScope = getFromScope(parentSelect);
+    ProcedureNamespace tableNs = new ProcedureNamespace(this,
+        fromScope, functionCall, selectItem);
+    tableNs.validateImpl(unknownType);
+    registerNamespace(null, null,
+        tableNs, false);
+    AliasNamespace aliasNs = new AliasNamespace(this,
+        selectItem, parentSelect);
+    aliasNs.validateImpl(unknownType);
+    registerNamespace(getSelectScope(parentSelect),
+        tableAlias, aliasNs, false);
+
+    // Create a table scope for table function
+    TableScope tableScope = new TableScope(fromScope, parentSelect);
+    if (fromScope instanceof ListScope) {
+      for (ScopeChild child : ((ListScope) fromScope).children) {
+        tableScope.addChild(child.namespace, child.name, child.nullable);
+      }
+    }
+    scopes.put(functionCall, tableScope);
+    // Associate the select with the table function
+    selectTableFunctions.put(parentSelect, functionCall);
+
```

Comment:
I'm skeptical about this member `selectTableFunctions`, the SqlValidator never has such mapping, it handles all the states to the "Namespace" and "scope", i'm wondering if we can move this mapping into each scope, because each select has a scope. Thus, we can keep the validator code clean.

Reply:
I think `selectTableFunctions `  is the same to the `whereScopes` or  `groupByScopes` in the `SqlValidatorImpl`.